### DISCUSSION TITLE
How to use environment variables  with docker and SSR?

### DISCUSSION THREAD

**Discussion by millievn** [vercel/next.js#15131] [Help] [answered]
I use caprover to deploy my project with docker and  I want to set different api address for dev/prod branches. 
But i find it's difficult to use environment variables and ssr when using docker.
Variables like  `process.env.api` will be replaced as string during build process. So when i use docker to run my project, environment variables won't work.
Any solution to use runtime environment variables?

**Comment by mulholo** [comment, marked as answer]
Have you checked out the docs on environment variables? https://nextjs.org/docs/basic-features/environment-variables

They got a big update recently.

**Reply by millievn** [reply]
I try it but it does not work.
`process.env.api` will be replaced automatically in build process. I know I can use dynamic environment  variables in [Next.js data fetching methods](https://nextjs.org/docs/basic-features/data-fetching) and [API routes](https://nextjs.org/docs/api-routes/introduction). But this means `process.env.api` must used in a page file like `/src/pages/blogs` or `/src/api/blogs`.
What i want is like this:

1. provide singleton client
```js
// /services/client.js

export default axios.create({
  baseURL: process.env.api,
});

// /pages/blogs.js

import client from '../services/client.js'

export async function getServerSideProps() {
  const res = await client.get(`https://.../data`)
  const data = await res.json()

  return { props: { data } }
}

```
2. tag a images
```bash
docker build -t my-blog .
```

Then we can run with different `process.env.api` without rebuild this project.

**Reply by mulholo** [reply]
> Then we can run with different process.env.api without rebuild this project.

Why would you want to change the api value inbetween builds? This is something you would only want to set once for your API and I imagine this would change very rarely.

**Reply by millievn** [reply]
> > Then we can run with different process.env.api without rebuild this project.
> 
> Why would you want to change the api value inbetween builds? This is something you would only want to set once for your API and I imagine this would change very rarely.

forget it. I use [runtime config](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration) the wrong way.
feel like a fool 😁

**Reply by houstonheat** [reply]
@qiwang97 Is there any chance of getting a code reference/example? Feel like a fool too 🌚

**Reply by millievn** [reply]
> @qiwang97 Is there any chance of getting a code reference/example? Feel like a fool too 🌚

See the [demo](https://github.com/qiwang97/nextjs-discussions-14966) 
The demo worked well in caprover. Please glance at [what seognil said](https://github.com/vercel/next.js/discussions/14966) if it not work.
